### PR TITLE
[SMTChecker] Create new auxiliary input callback and use as SMT solver

### PR DESCRIPTION
**Pull request by leonardoalt**
New version of #5675

Depends on #7768.

### CODE REVIEW

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `e0ade160591a2804a6be8c9cd22f3ed4a1dd7ff4`:
```
Coding style error:
 libsolidity/formal/SMTLib2Interface.cpp:221: auto result = m_smtCallback(_input);

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `98845cda464b28cfd280c0d61493d99bc6329644`:
```
Coding style error:
 libsolidity/formal/SMTLib2Interface.cpp:221: auto result = m_smtCallback(_input);

```
Please check that your changes are working as intended.

**Comment by leonardoalt** [comment]
Decided during call:

- There will be only one callback.
- This callback takes 3 arguments: context, kind, data. This PR only implements the kind part.

File: libsolidity/interface/ReadFile.h [hunk 1]
```diff
@@ -37,8 +37,21 @@ class ReadCallback: boost::noncopyable
...
+	{
+		ReadFile,
+		SMTQuery
+	};
+
+	static std::string kindString(Kind _kind)
+	{
+		if (_kind == Kind::SMTQuery)
+			return "smt-query";
+		return "source";
```
Comment (axic) [inline libsolidity/interface/ReadFile.h]: Why not a switch statement which would catch unmatched enum cases?

**Comment by leonardoalt** [comment]
This still needs https://github.com/ethereum/solidity/pull/7452 and https://github.com/ethereum/solidity/pull/7453 to be merged, and develop -> develop_060.

File: libsolidity/interface/ReadFile.h (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/interface/ReadFile.h]: Fixed

**Comment by leonardoalt** [comment]
The solc-js tests finally pass here using `smtCheckerTests` via `CVC4`. In the future, when `CHCSmtLib2Interface` is written, we need to switch to Z3 >= 4.8.5.

File: Changelog.md [hunk 2]
```diff
@@ -3,6 +3,7 @@
 Breaking changes:
  * ABI: remove the deprecated ``constant`` and ``payable`` fields.
  * ABI: the ``type`` field is now required and no longer specified to default to ``function``.
+ * C Interface: the provided callback now takes two parameters, kind and data. The callback can then be used for multiple purposes, such has file imports and SMT queries.
```
Comment (axic) [inline Changelog.md]: "such has" -> "such as"

File: libsolc/libsolc.cpp [hunk 3]
```diff
@@ -43,17 +43,17 @@ ReadCallback::Callback wrapReadCallback(CStyleReadFileCallback _readCallback = n
...
 			char* error_c = nullptr;
-			_readCallback(_path.c_str(), &contents_c, &error_c);
+			_readCallback(_kind.c_str(), _data.c_str(), &contents_c, &error_c);
 			ReadCallback::Result result;
 			result.success = true;
 			if (!contents_c && !error_c)
 			{
 				result.success = false;
-				result.responseOrErrorMessage = "File not found.";
+				result.responseOrErrorMessage = "Callback not found.";
```
Comment (axic) [inline libsolc/libsolc.cpp]: Hm, not sure this is the right way to do it. Perhaps we should let empty `contents_c` pass through as valid (it could be, depending on the kind) and force the callback to return `File not found` if that's the case.

File: Changelog.md (hunk 2 above)
Comment (leonardoalt) [reply Changelog.md]: updated

File: libsolc/libsolc.cpp (hunk 3 above)
Comment (leonardoalt) [reply libsolc/libsolc.cpp]: But isn't `File not found` relevant only for the import callback?

Comment (axic) [reply libsolc/libsolc.cpp]: Exactly, and this is a generic piece of code.

Comment (axic) [reply libsolc/libsolc.cpp]: Ah OK we can mandate to return an empty string (still an allocated pointer).

Comment (leonardoalt) [reply libsolc/libsolc.cpp]: Wait, what case would that be? If content is valid but empty? Can you write the new rules here?

**Comment by axic** [comment]
Wow this passes all tests now!

File: scripts/test_emscripten.sh [hunk 4]
```diff
@@ -40,3 +40,5 @@ VERSION=$("$REPO_ROOT"/scripts/get_version.sh)
 
 echo "Running solcjs tests...."
 "$REPO_ROOT/test/externalTests/solc-js/solc-js.sh" "$SOLJSON" "$VERSION"
+echo "Running external tests...."
+"$REPO_ROOT/test/externalTests.sh" "$SOLJSON"
```
Comment (axic) [inline scripts/test_emscripten.sh]: Why is this needed?

File: solc/CommandLineInterface.cpp [hunk 5]
```diff
@@ -807,10 +807,15 @@ Allowed options)",
 
 bool CommandLineInterface::processInput()
 {
-	ReadCallback::Callback fileReader = [this](string const& _path)
+	ReadCallback::Callback fileReader = [this](string const& _kind, string const& _path)
 	{
 		try
 		{
+			if (_kind != ReadCallback::kindString(ReadCallback::Kind::ReadFile))
```
Comment (axic) [inline solc/CommandLineInterface.cpp:855]: What case can this happen? Why not return an error message and not an exception?

File: libsolidity/interface/ReadFile.h [hunk 6]
```diff
@@ -37,8 +39,27 @@ class ReadCallback: boost::noncopyable
...
+	};
+
+	static std::string kindString(Kind _kind)
+	{
+		switch (_kind)
+		{
+		case Kind::ReadFile:
+			return "source";
+		case Kind::SMTQuery:
+			return "smt-query";
```
Comment (axic) [inline libsolidity/interface/ReadFile.h:55]: Not fully sure having these two names, but have no other proposal.

**Comment by axic** [comment]
@leonardoalt apart from these simple questions this looks good!

File: scripts/test_emscripten.sh (hunk 4 above)
Comment (leonardoalt) [reply scripts/test_emscripten.sh]: Not sure, will remove

File: solc/CommandLineInterface.cpp (hunk 5 above)
Comment (leonardoalt) [reply solc/CommandLineInterface.cpp:855]: Only internally in case someone coded it wrong. Same idea as `solAssert`

File: test/externalTests/solc-js/solc-js.sh [hunk 7]
```diff
@@ -35,7 +35,7 @@ function solcjs_test
     SOLCJS_INPUT_DIR="$TEST_DIR"/test/externalTests/solc-js
 
     # set up solc-js on the branch specified
-    setup master_060
+    setup smt_callback
```
Comment (leonardoalt) [inline test/externalTests/solc-js/solc-js.sh]: @axic Should we first merge the `solc-js` PR and then change this back to `master` actually?

Comment (axic) [reply test/externalTests/solc-js/solc-js.sh]: Yes, that's the plan.

File: libsolidity/interface/ReadFile.h (hunk 6 above)
Comment (leonardoalt) [reply libsolidity/interface/ReadFile.h:55]: Do you mean having the names at all or these names specifically?

File: scripts/test_emscripten.sh (hunk 4 above)
Comment (leonardoalt) [reply scripts/test_emscripten.sh]: Probably conflict resolving mistake, removed.

File: libsolc/libsolc.cpp (hunk 3 above)
Comment (leonardoalt) [reply libsolc/libsolc.cpp]: @axic any update required here? I'm still not sure I understood it

**Comment by leonardoalt** [comment]
Had two questions back. Once those are resolved and https://github.com/ethereum/solc-js/pull/382 is merged, we can point to `solc-js/master` again and squash.

File: libsolidity/interface/ReadFile.h (hunk 6 above)
Comment (axic) [reply libsolidity/interface/ReadFile.h:55]: These specific names.

File: libsolc/libsolc.cpp (hunk 3 above)
Comment (axic) [reply libsolc/libsolc.cpp]: I'm confused myself.

**Comment by axic** [comment]
Retriggered tests to see if it works properly with the new version of the solc-js PR.

**Comment by leonardoalt** [comment]
Error: `TypeError: solc.compileStandardWrapper is not a function`

**Comment by leonardoalt** [comment]
I'll try to figure locally what's happening

**Comment by leonardoalt** [comment]
Ah, it's actually the DAO tests that were moved to this repo

**Comment by leonardoalt** [comment]
Fixed `determinism.js` and rebased

**Comment by leonardoalt** [comment]
@axic  Tests passing ;)

File: test/externalTests/solc-js/solc-js.sh (hunk 7 above)
Comment (axic) [inline test/externalTests/solc-js/solc-js.sh]: Needs to be changed back to `master_060`

File: libsolc/libsolc.cpp (hunk 3 above)
Comment (axic) [reply libsolc/libsolc.cpp]: Moved this to #7769.

File: solc/CommandLineInterface.cpp [hunk 8]
```diff
@@ -848,10 +848,15 @@ Allowed options)",
...
+				BOOST_THROW_EXCEPTION(InternalCompilerError() << errinfo_comment(
```
Comment (axic) [inline solc/CommandLineInterface.cpp:856]: Actually can use `solAssert` here, that generates an ICE.

Comment (axic) [reply solc/CommandLineInterface.cpp:856]: Never mind, the CLI doesn't use `solAssert`s.

**Comment by axic** [comment]
Rebased. Ready for merging.
